File: ruby/isomorphic_strings.rb
# frozen_string_literal: true

# @param {String} s
# @param {String} t
# @return {Boolean}

def is_isomorphic(s, t)

  letter_map = {}
  reverse_map = {}

  s.chars.each_with_index do |char_s, i|
    char_t = t[i]

    return false if letter_map[char_s] && letter_map[char_s] != char_t
    return false if reverse_map[char_t] && reverse_map[char_t] != char_s

    letter_map[char_s] = char_t
    reverse_map[char_t] = char_s
  end

  true
end

File: typescript/isomorphicStrings.ts
function isIsomorphic(s: string, t: string): boolean {
  let letterMap: { [key: string]: string } = {};
  let reverseMap: { [key: string]: string } = {};

  for (let i = 0; i < s.length; i++) {
    let charS = s[i];
    let charT = t[i];

    if (letterMap[charS] && letterMap[charS] != charT) {
      return false;
    }

    if (reverseMap[charT] && reverseMap[charT] != charS) {
      return false;
    }

    letterMap[charS] = charT;
    reverseMap[charT] = charS;
  }

  return true;
}
